src/Stratis.Bitcoin/EventBus/InMemoryEventBus.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Stratis.Bitcoin.EventBus
+{
+    public class InMemoryEventBus : IEventBus
+    {
+        /// <summary>
+        /// The subscriptions stored by EventType
+        /// </summary>
+        private readonly Dictionary<Type, List<ISubscription>> subscriptions;
+
+        /// <summary>
+        /// The subscriptions lock to prevent race condition during publishing
+        /// </summary>
+        private static readonly object SubscriptionsLock = new object();
+
+        public InMemoryEventBus()
+        {
+            this.subscriptions = new Dictionary<Type, List<ISubscription>>();
+        }
+
+        /// <inheritdoc />
+        public SubscriptionToken Subscribe<TEventBase>(Action<TEventBase> action) where TEventBase : EventBase
+        {
+            if (action == null)
+                throw new ArgumentNullException(nameof(action));
+
+            lock (SubscriptionsLock)
+            {
+                if (!this.subscriptions.ContainsKey(typeof(TEventBase)))
+                    this.subscriptions.Add(typeof(TEventBase), new List<ISubscription>());
+
+                var token = new SubscriptionToken(this, typeof(TEventBase));
+                this.subscriptions[typeof(TEventBase)].Add(new Subscription<TEventBase>(action, token));
+                return token;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Unsubscribe(SubscriptionToken token)
+        {
+            if (token == null)
+                throw new ArgumentNullException(nameof(token));
+
+            lock (SubscriptionsLock)
+            {
+                if (this.subscriptions.ContainsKey(token.EventType))
+                {
+                    var allSubscriptions = this.subscriptions[token.EventType];
+
+                    var subscriptionToRemove = allSubscriptions.FirstOrDefault(sub => sub.SubscriptionToken.Token == token.Token);
+                    if (subscriptionToRemove != null)
+                        this.subscriptions[token.EventType].Remove(subscriptionToRemove);
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public void Publish<TEventBase>(TEventBase eventItem) where TEventBase : EventBase
+        {
+            if (eventItem == null)
+                throw new ArgumentNullException(nameof(eventItem));
+
+            // Assigns an unique id to the event.
+            eventItem.CorrelationId = Guid.NewGuid();
```

Comment:
Can this be assigned in the `EventBase` constructor? Would also prevent an issue where `EventBase.ToString` can be called while the CorrelationId is `null`.

Reply:
Yes, I agree. Event itself should create a correlation id. It would also help with deduping the same events/commands being sent twice.

Reply:
yes I agree, actually it was here because previously it was a progressive (long) value (that couldn't be known by EventBase class) and then I refactored it to be a correlation Id with a Guid but Guid can of course be instantiated inside the EventBase class.